core/src/main/scala/org/apache/spark/ui/HttpSecurityFilter.scala

```diff
@@ -0,0 +1,116 @@
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.ui
+
+import javax.servlet._
+import javax.servlet.http.{HttpServletRequest, HttpServletResponse}
+
+import scala.collection.JavaConverters._
+
+import org.apache.commons.lang3.StringEscapeUtils
+
+import org.apache.spark.{SecurityManager, SparkConf}
+import org.apache.spark.internal.config._
+
+/**
+ * A servlet filter that implements HTTP security features. The following actions are taken
+ * for every request:
+ *
+ * - perform access control of authenticated requests.
+ * - check request data for disallowed content (e.g. things that could be used to create XSS
+ *   attacks).
+ * - set response headers to prevent certain kinds of attacks.
+ */
+private class HttpSecurityFilter(
+    conf: SparkConf,
+    securityMgr: SecurityManager) extends Filter {
+
+  private val NEWLINE_AND_SINGLE_QUOTE_REGEX = raw"(?i)(\r\n|\n|\r|%0D%0A|%0A|%0D|'|%27)".r
+
+  override def destroy(): Unit = { }
+
+  override def init(config: FilterConfig): Unit = { }
+
+  override def doFilter(req: ServletRequest, res: ServletResponse, chain: FilterChain): Unit = {
+    val hreq = req.asInstanceOf[HttpServletRequest]
+    val hres = res.asInstanceOf[HttpServletResponse]
+    hres.setHeader("Cache-Control", "no-cache, no-store, must-revalidate")
+
+    if (!securityMgr.checkUIViewPermissions(hreq.getRemoteUser())) {
+      hres.sendError(HttpServletResponse.SC_FORBIDDEN,
+        "User is not authorized to access this page.")
+      return
+    }
+
+    // Check if any disallowed content is in the incoming headers or parameters. This filters
+    // out content that could be used for XSS attacks from even making it to the UI handlers.
+    try {
+      hreq.getHeaderNames().asScala.foreach { k =>
+        require(isSafe(k), "Request header name contains disallowed content.")
+        hreq.getHeaders(k).asScala.foreach { v =>
+          require(isSafe(v), s"Header value for $k contains disallowed content.")
+        }
+      }
+
+      hreq.getParameterMap().asScala.foreach { case (k, values) =>
+        require(isSafe(k), "Parameter name contains disallowed content.")
+        values.foreach { v =>
+          require(isSafe(v), s"Parameter value for $k contains disallowed content.")
+        }
+      }
+    } catch {
+      case e: IllegalArgumentException =>
```

Comment:
If other type of exception is coming, will there be a response sent to the client?

Reply:
You'll get an error 500.